### ISSUE TITLE
🐞 Exclude pattern seems inconsistent

### ISSUE THREAD

**Issue by benjaminstrasser** [dagger/dagger#3791]
### What is the issue?

When trying out 4 different approaches to excluding a file(test.txt)/directory(build) from a directory I am getting unexpected results.

In short, dir behaves as follows:

test.txt is excluded
```
workdir := client.Host().Workdir()

dir := client.Directory().WithDirectory(".", workdir, dagger.DirectoryWithDirectoryOpts{
	Exclude: []string{"test.txt"},
})
```

build is excluded
```
dir := client.Host().Workdir(dagger.HostWorkdirOpts{
	Exclude: []string{"build"},
})
```

build exists but its contents are excluded
```
workdir := client.Host().Workdir()

dir := client.Directory().WithDirectory(".", workdir, dagger.DirectoryWithDirectoryOpts{
	Exclude: []string{"build/*"},
})
```

This causes and error
```
workdir := client.Host().Workdir()

dir := client.Directory().WithDirectory(".", workdir, dagger.DirectoryWithDirectoryOpts{
	Exclude: []string{"build"},
})
```

### Log output

```
$ mage Fails
#1 resolve image config for docker.io/library/alpine:latest
#1 DONE 0.5s

#2 mkfile /Dockerfile
#2 DONE 0.0s

#3 mkfile /main.go
#3 CACHED

#4 [internal] load metadata for docker.io/library/golang:1.18.2-alpine

#4 DONE 0.2s

#5 [build 1/6] FROM docker.io/library/golang:1.18.2-alpine@sha256:4795c5d21f01e0777707ada02408debe77fe31848be97cf9fa8a1462da78d949
#5 resolve docker.io/library/golang:1.18.2-alpine@sha256:4795c5d21f01e0777707ada02408debe77fe31848be97cf9fa8a1462da78d949 0.0s done
#5 DONE 0.0s

#6 local://C:\Users\bstrasser\Documents\Projects\Fun\reproduce-bug\hello-monorepo
#6 transferring C:\Users\bstrasser\Documents\Projects\Fun\reproduce-bug\hello-monorepo: 4.50kB 0.0s done
#6 DONE 0.0s

#7 copy / /
#7 CACHED

#8 mkdir /meta
#8 CACHED

#9 [build 2/6] WORKDIR /src
#9 CACHED

#10 [build 6/6] RUN CGO_ENABLED=0 go build -o /_shim -ldflags '-s -d -w' .
#10 CACHED

#11 [build 4/6] RUN go mod init go.dagger.io/dagger/shim/cmd
#11 CACHED

#12 [build 5/6] COPY . .
#12 CACHED

#13 [build 3/6] RUN apk add --no-cache file git
#13 CACHED

#14 [stage-1 1/1] COPY --from=build /_shim /_shim
#14 CACHED

#15 copy / /
#15 ERROR: failed to copy file info: failed to chown /var/lib/buildkit/runc-overlayfs/snapshots/snapshots/730/fs/build: lchown /var/lib/buildkit/runc-overlayfs/snapshots/snapshots/730/fs/build: no such file or directory    

#16 docker-image://docker.io/library/alpine:latest
#16 resolve docker.io/library/alpine:latest 0.0s done
#16 CANCELED
------
 > copy / /:
------

```

### Steps to reproduce

[Repo for reproduction](https://github.com/benjaminstrasser/hello-monorepo/tree/reproduce-bug)

### Dagger version

dagger 0.2.32 (624e5bb94) windows/amd64

### OS version

Windows 10

**Comment by benjaminstrasser** [comment]
Reproducible with `dagger 0.2.36 (ea275a3ba) windows/amd64`

**Comment by benjaminstrasser** [comment]
Tagging #3802

**Comment by aluzzardi** [comment]
Thank you @benjaminstrasser! We re-use the same code as `.dockerignore` and inherit the same patterns.

Could you tell me your expected results for these?

AFAIK -- `test.txt` and `build` exclude **any** file/directory named as such.

`/test.txt` would exclude that one specific file.

/cc @sipsma @vito

**Comment by benjaminstrasser** [comment]
### Case 1 ✅
Actual result:       test.txt is excluded
Expected result:  test.txt is excluded
> ```
> workdir := client.Host().Workdir()
> 
> dir := client.Directory().WithDirectory(".", workdir, dagger.DirectoryWithDirectoryOpts{
> 	Exclude: []string{"test.txt"},
> })
> ```


### Case 2 ✅(test case added in #3802)
Actual Result:      build is excluded
Expected Result:  build is excluded
> ```
> dir := client.Host().Workdir(dagger.HostWorkdirOpts{
> 	Exclude: []string{"build"},
> })
> ```


### Case 3 ✅
Actual Result:      build exists but its contents are excluded 
Expected Result:  build exists but its contents are excluded 
> ```
> workdir := client.Host().Workdir()
> 
> dir := client.Directory().WithDirectory(".", workdir, dagger.DirectoryWithDirectoryOpts{
> 	Exclude: []string{"build/*"},
> })
> ```


### Case 4 ❌(test case added in #3802)
Actual Result:      error is thrown
Expected Result:  build is excluded
 
> ```
> workdir := client.Host().Workdir()
> 
> dir := client.Directory().WithDirectory(".", workdir, dagger.DirectoryWithDirectoryOpts{
> 	Exclude: []string{"build"},
> })
> ```


In #3802 test cases for case 2 and case 4 were added to illustrate actual and expected behavior.

**Comment by vito** [comment]
Relaying https://github.com/dagger/dagger/pull/3802?notification_referrer_id=NT_kwDNB1ivNDgyMjYzMTI3OToxODgw#issuecomment-1335730599 - this is an issue in upstream Buildkit.

**Comment by grouville** [comment]
Closed by #3802
